src/metabase/search/filter.clj

```diff
@@ -81,10 +81,26 @@
 (defn personal-collections-where-clause
   "Build a clause limiting the entries to those (not) within or within personal collections, if relevant.
   WARNING: this method queries the appdb, and its approach will get very slow when there are many users!"
-  [{filter-type :filter-items-in-personal-collection} collection-id-col]
-  (when filter-type
-    (let [parent-ids     (t2/select-pks-vec :model/Collection :personal_owner_id [:not= nil])
-          child-patterns (for [id parent-ids] (format "/%d/%%" id))]
+  [{filter-type :filter-items-in-personal-collection :keys [current-user-id] :as search-ctx} collection-id-col]
+  (case (or filter-type "all")
+    "all" nil
+
+    "only-mine"
+    [:or
+     [:= :collection.personal_owner_id current-user-id]
+     (into [:or]
+           (let [your-collection-ids (t2/select-pks-vec :model/Collection :personal_owner_id [:= current-user-id])
+                 child-patterns      (for [id your-collection-ids] (format "/%d/%%" id))]
+             (for [p child-patterns] [:like :collection.location p])))]
```

Comment:
it iterates through all your collections, do you think that maybe it can have `=` here? Also can we make it through the index/filters maybe, like put a collection owner id in filters?

Reply:
Putting the collection owner id in the index is going to be a bit tricky, I'm going to leave that for later this week / next week. It'll be good to do that the same time as putting in the effective path.

As for putting it in the filter, yeah I'm not sure why I didn't do that. I think it can work, will give it a go.

As for using an `=`, I don't think so. The non-personal sub-collections of personal collections can be nested arbitrarily deep. It would be nice if we marked all the sub-collections with the personal_owner_id as well, but we don't.